### ISSUE TITLE
Make uploading of build trees configurable

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#566]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/566)
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Aug 8, 2018, 16:58

## Summary

The build tree should be optional in the artifact.  When build elements are known not to be trusted for incremental builds, uploading the build tree is very wasteful.  In scenarios where the vast majority of builds are going to be clean builds, the uploading of build trees is wasteful as well.
The storage requirements of build trees are typically a multiple of the storage requirements without the build tree.  Even independent expiry of the build tree, should be an option to trade space vs time.  Or rather reclaim space when there is no use for the build tree any longer.

I missed this in #21:
> We do not want any project.conf variable related to this at all if it can be avoided
> - The cost of considering that an artifact might, or might not have a cached build tree available is going to be increasingly expensive, this is a distinction we don't want to have to make.

I do think we do want that for reasons above.  And I do think that the client needs to gracefully deal with the fact that a build tree may not be present for an artifact.

I can see the following options be valid either in project.conf, user configuration or element configuration:
- failure: only upload the build tree on build failures
- always: always upload the build tree
- never: never upload the build tree

That way you can configure the project to only upload the build tree in case of failure, and then e.g. in webkit.bst configure that specific element to upload its build tree.

Edit, added by Laurence, 21 Nov: see related Mailing List discussion [here](https://mail.gnome.org/archives/buildstream-list/2018-November/msg00004.html)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Aug 9, 2018, 07:10

Hi Sander,

While I do agree that this can be interesting, we have just introduced the functionality on master and we need to ensure that developers are using it on master, improving it, and making the feature is useful before we start down the path of disabling it.

Orthogonally, I should also point out that so far, under no circumstances at all is an incremental build trustable, as we have no way to ensure the inputs are identical or if the result would be the same. If we alter the cache key calculation for it, we would again have no way of reproducing the same artifact twice.

We should also take into account that deduplication on the artifact server + reproducible builds, should mitigate both the storage and upload time cost of this significantly. With SourceCache on the horizon, assuming this will be done in the same CAS as the artifact cache, we also gain at local caching time. We should be better informed of just how exponential the cost is going to be with all the intended pieces in place before talking about disabling it.


### Most importantly

The build tree caching featureset needs to be useful to everyone, not only a subset of users/scenarios/elements where it is acceptable to upload/download/store the build trees all the time, that would obviously be unacceptable.

For build tree caching to be useful, we have to address #494 *first*. Any build tree disabling options we might add will not cause #494 to stop blocking the build tree caching featureset from inclusion in 1.4.

The risk that we fail to make the feature work well in scenarios where build servers do most of the computing and most/all of the uploading to a shared cache anyway (the most likely scenario so far), and where users only need to optionally download the build tree on demand when needed, is too high to start adding options to disable it at this time. A statement that *"if the feature doesnt work for you, then you can disable it"*, can not be used as an excuse to avoid the effort of making it work well for everyone.

Once we have properly addressed #494, we can re-assess what exactly the costs are in practice, and yes it will probably make sense to consider project.conf/element/user config options at that time, but not before.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Aug 9, 2018, 10:32

Hi Tristan,

> While I do agree that this can be interesting, we have just introduced the functionality on master and we need to ensure that developers are using it on master, improving it, and making the feature is useful before we start down the path of disabling it.

I think the disconnect here that me filing the issue is seen as "this needs to happen now otherwise the feature will be useless" vs "this needs to be considered to make the feature useful long term and in all use cases".

> Orthogonally, I should also point out that so far, under no circumstances at all is an incremental build trustable, as we have no way to ensure the inputs are identical or if the result would be the same.

That incremental builds have a different trust level in the context of BuildStream, that I fully agree with.  However, I was referring to the fact that some builds simply don't do the right thing when run incrementally.  Why would we want to store the build tree for these in the common case?

> We should also take into account that deduplication on the artifact server + reproducible builds, should mitigate both the storage and upload time cost of this significantly. With SourceCache on the horizon, assuming this will be done in the same CAS as the artifact cache, we also gain at local caching time. We should be better informed of just how exponential the cost is going to be with all the intended pieces in place before talking about disabling it.

You really lost me here.  I think the key part where we are differing is "reproducible builds".  With BuildStream we have reproducible inputs and therefor say we can reuse the outputs.  Fully reproducible builds where every element produces the exact same output is likely a goal for many, but can't be considered the current state of the world.  As such the deduplication may be overestimated at this point for artifacts.  It won't be zero but it won't be perfect either.  Disk space and network bandwidth should both be considered; neither is unlimited.
I don't quite see how SourceCache would factor into this, as it has nothing to do with build tree caching.
That we need actual data I agree with.  On the flip side of your argument, having a feature that we don't know the impact of, and consciously witholding a kill switch, I would not consider a best practice. 

> ### Most importantly

> The build tree caching featureset needs to be useful to everyone, not only a subset of users/scenarios/elements where it is acceptable to upload/download/store the build trees all the time, that would obviously be unacceptable.

+1.

> For build tree caching to be useful, we have to address #494 _first_. Any build tree disabling options we might add will not cause #494 to stop blocking the build tree caching featureset from inclusion in 1.4.

Agreed that #494 is a must have.

> The risk that we fail to make the feature work well in scenarios where build servers do most of the computing and most/all of the uploading to a shared cache anyway (the most likely scenario so far), and where users only need to optionally download the build tree on demand when needed, is too high to start adding options to disable it at this time. A statement that _"if the feature doesnt work for you, then you can disable it"_, can not be used as an excuse to avoid the effort of making it work well for everyone.

I think that may be dismissing the scenarios where it's not build servers doing most of the computing and most/all of the uploading.  I feel that your twisting my words a bit here?

> Once we have properly addressed #494, we can re-assess what exactly the costs are in practice, and yes it will probably make sense to consider project.conf/element/user config options at that time, but not before.

I have not stated anything regarding timing of this issue, nor ordering with respect to other issues.  If we can agree this is an enhancement, then I think we're on the same page?

Cheers,

Sander

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Aug 9, 2018, 10:55

Sander, thanks a lot for clarifying.

I am sincerely sorry to act defensively on this, landing the build tree caching opens up a lot of possibilities that I need to be careful of - this was a knee-jerk reaction on my part.

Yes we are certainly on the same page here - I do want to see this used a lot before introducing an escape route, but agree that we should design one at some point before freezing things in 1.4.

> You really lost me here. I think the key part where we are differing is "reproducible builds". With BuildStream we have reproducible inputs and therefor say we can reuse the outputs. Fully reproducible builds where every element produces the exact same output is likely a goal for many, but can't be considered the current state of the world. As such the deduplication may be overestimated at this point for artifacts. It won't be zero but it won't be perfect either. Disk space and network bandwidth should both be considered; neither is unlimited. I don't quite see how SourceCache would factor into this, as it has nothing to do with build tree caching. That we need actual data I agree with. On the flip side of your argument, having a feature that we don't know the impact of, and consciously witholding a kill switch, I would not consider a best practice.

To clarify what I meant with the SourceCache, if I understand it correctly; it will be using the same merkle trees and resides in the same underlying CAS which the ArtifactCache uses, as such I would expect that at least while caching the build tree itself; everything which is source files need not be added physically to the cache (this does not optimize everything away at cache time, but it buys something).

For reproducibility, I do think that we can get very close just by controlling the determinism of the build environment (this has proven mostly true in the Flatpak builds on flathub, but we should verify the numbers). Of course you are right that that does not cover everything, but if it covers 90% of data, that is a huge reduction in what we are caching and potentially transporting.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Aug 10, 2018, 16:17

> I am sincerely sorry to act defensively on this, landing the build tree caching opens up a lot of possibilities that I need to be careful of - this was a knee-jerk reaction on my part.

No worries.  Thanks for your candor.

> To clarify what I meant with the SourceCache, if I understand it correctly; it will be using the same merkle trees and resides in the same underlying CAS which the ArtifactCache uses, as such I would expect that at least while caching the build tree itself; everything which is source files need not be added physically to the cache (this does not optimize everything away at cache time, but it buys something).

You are not wrong per se there.  However, regardless of SourceCache, you can already discount the source files, because they would already be in CAS due to Remote Execution of the build (as part of the input).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Nov 21, 2018, 12:47

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Nov 23, 2018, 15:10

Hi all, I'm currently trying to assess where we currently are with the underlying support for this, especially since most of https://gitlab.com/BuildStream/buildstream/issues/494 has landed. 

The last remaining task is https://gitlab.com/BuildStream/buildstream/issues/774, which is definitely useful when considering handling partial artifacts in a more seamless manner for the user. I've taken time to try and digest the mail thread about this issue, currently ending in https://mail.gnome.org/archives/buildstream-list/2018-November/msg00025.html which has a relatable point 

>  We need to design some abstraction to handle failure modes for
    partial artifacts, possibly at the ArtifactCache level where
    components might ask the ArtifactCache for a specific component
    and the program is required to fail in a specific way

As it stands, https://gitlab.com/BuildStream/buildstream/merge_requests/786 had to handle this consideration all be it with a rudimentary clause. It adds api `contains_subdir_artifact()` which is used to query the element digest for a given subdir string (buildtree) checking that it's not a dangling ref. In it's current naive implementation, this limits numerous assumptions where the buildtree is required. This includes attempting to upload a partial artifact, staging a buildtree into the sandbox, and coming in https://gitlab.com/BuildStream/buildstream/issues/311 checking-out the buildtree into workspaces. This requires querying the CAS each time, updating `Element._update_state()` to track this would be much more efficient, but I must admit I still need to grok it completely. 

If https://gitlab.com/BuildStream/buildstream/issues/774 leads to being able to gracefully spawn attempts at completing the buildtree when required without the user being specific, then I think this will lower the surface for failure. However I'm not sure if this would warrant a new user option, i.e if not available locally or any remote drop back to some default behaviour, or purposely fail cleanly.

The bulk of the above is related to pulling partial artifacts, where we know currently know that if the artifact server has the artifact digest, it has the all of the content available. An initial thought could be 'what if a partial pull job for a buildtree is attempted but the upstream artifact is missing it'. This is not currently tested for given the previous assumption given. The current logic of cascache.py `pull()` through the use of `self._fetch_subdir(remote, tree, subdir)` allows the respective element's specific subdir to be pulled, again with the assumption that because we've found a remote with the elements ref, then it will have the subdir. An exception used to only get thrown if the top level element ref is not found. We'd need to consider handling this, there's obviously numerous ways to do this and I think it should be done with the least amount of friction on the user side. We also obviously have pull_tree() which takes a specific digest (could be buildtree) and handles pulling it, along with the grpc exception handler.

There are my initial thoughts, highlighting what I believe is the current situation going forward to support this issue.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Nov 23, 2018, 16:05

One side note, I also feel like the ability to upload partial artifacts (if there's no rejection configuration added serverside) should be only allowed via user configuration.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Nov 23, 2018, 17:14

I maintain the stance that an artifact without a build tree is not incomplete, as it should be optional.  Artifacts that had no build trees previously are not incomplete.  And they were still very useful to depend on without them.

When you say:
> One side note, I also feel like the ability to upload partial artifacts (if there's no rejection configuration added serverside) should be only allowed via user configuration.

Do you imply that it shouldn't be possible using project configuration?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 26, 2018, 05:05

> I maintain the stance that an artifact without a build tree is not incomplete, as it should be optional. Artifacts that had no build trees previously are not incomplete. And they were still very useful to depend on without them.

I really want to agree with you [[Gitlab user @sstriker]](https://gitlab.com/sstriker).

Cached build trees were only thought to have relevance for optional features, like incremental builds for an initially opened workspace, debugging of failed builds, or potential weird (less trusted) build modes which are incremental across the board.

But now that the build trees exist, they have unfortunately went ahead and tangled themselves into a [potential core feature](https://mail.gnome.org/archives/buildstream-list/2018-November/msg00042.html), for the sake of the testing plans, this is unfortunately going to require serious thought process.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 26, 2018, 05:17

[[Gitlab user @tpollard]](https://gitlab.com/tpollard), here is a separate response from the difficult elephant in the room pointed out to [[Gitlab user @sstriker]](https://gitlab.com/sstriker) above (I think that regardless of testing strategies, we still want this optionality)...

You are saying that whether or not you download build trees is a user decision, thus this is user configuration; I agree with you 100% - however I am not convinced that the project should not be allowed to provide a recommendation.

To be clear, if the project.conf received any option, this would be a recommendation that a project makes which only overrides the defaults of a user configuration (we've been calling these *"project recommendations"* so far, maybe better to make that terminology *"a thing*") - there should be no confusion between this and project data; the option is still user configuration, and the project can only provide a recommendation.

[[Gitlab user @sstriker]](https://gitlab.com/sstriker), I think in this case the best way forward is always to *"not do something"* until it's asked for. I'm don't feel strongly about denying the `project.conf` the opportunity to provide a recommendation on this configuration (as it does in a few other cases for other user configurations), but I *do* feel strongly about doing as little as possible at any given time especially when it comes to API surface (don't do it until it's proven that someone wants it).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Nov 27, 2018, 12:33

>  if there's no rejection configuration added serverside

Is this something that we need to consider when adding opt-in support for pushing partial artifacts? Should the server have knowledge of what 'defines' a 'complete' artifact, or should that be left to the client?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 27, 2018, 12:53

> Is this something that we need to consider when adding opt-in support for pushing partial artifacts? Should the server have knowledge of what 'defines' a 'complete' artifact, or should that be left to the client?

Thus far the server has forever been *dumb* with regards to what it stores, which is rather a good thing.

There are 3 classes of configuration essentially:
* Project data, this is the input of BuildStream, the *"source code"* if you will. This usually has an impact on the output of a project, it can set defaults for build instructions, environment variables, sandbox configurations and the like. One could say that project data is declared by the `project.conf` and the sum of the element `.bst` files owned by the project.
* Configurations dictated by users, these are generally preferences, and should never effect the output of a build
  * Note that in this case, excepting for the [snag I pointed out above](https://mail.gnome.org/archives/buildstream-list/2018-November/msg00042.html), build trees have always been considered things that are not needed, but can be helpful for certain features, so we have not been considering them as *"build output"*
* Project recommended configurations: This is when the project recommends a given user preference to the users, it overrides the default user preference but can be overridden by the user's preference or command line options
  * This initially presented itself in the form of artifact cache servers, when we sat down at GUADEC 2017 in Manchester and discussed that it would be convenient for working groups to inherit the artifact server address from the `project.conf`, so people don't have to set it up themselves

To add to the above, it is totally fair that multiple working groups share the same artifact server, but may be storing artifacts from different projects, those projects might have different *"project recommendations"* for user configurations, such as whether users of this project should overall be uploading build trees by default.

As I already stated in a previous comment, I don't feel particularly strongly about allowing such a configuration to exist if it is clearly asked for - I am only against adding any API surface which has never really been asked for, because API additions are irreversible once added in a stable release.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 29, 2018, 12:13

~~Honestly regardless of the comments above about why I think we currently *do* have the correct default, I'm very inclined to reject this solely based on the fact that it is a very invasively breaking change.~~

~~Literally any project which uses any git sources which have submodules would break as a result of this change - while I know for a fact that a known project like `gnome-build-meta` has *several* (probably even close to 10, definitely more than 5) elements which have submodules which the authors previously did not have to care about - it stands to reason that there is a very good chance this breaks *any* projects that we dont know about and track closely - it's a very unfriendly change.~~

Update: Crap - So sorry, the above comment landed on the completely wrong issue, this was meant for #783

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Nov 29, 2018, 16:19

Could it make sense to add a config key `push-buildtrees` to the artifact server configuration dict? That's already supported as both user configuration and project recommendation and having this configuration per specified artifact server sounds reasonable to me (basic `push` is configured in the same place).

I would like to avoid a server-side configuration for this. The idea of `ReferenceStorage` is to be agnostic of the contents and we will use the same mechanism for source cache, which obviously won't have buildtrees.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2018, 09:46

[[Gitlab user @juergbi]](https://gitlab.com/juergbi) I'm not sure it really makes sense (I would also like to avoid server side configuration that has any knowledge of the nature or policies revolving around the content).

That seems to make some presumptions that the artifact server/storage is coupled with a given project or organization.

While I do think that the user is best served if the decision of uploading build trees is in the users hands, and the project only makes a recommendation, I think that the farthest we would want to take this is to introduce a concept of a policy dictated to the user by the project.

In other words, it's easily conceivable that various projects with different policies share the same artifact server.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Nov 30, 2018, 09:53

> In other words, it's easily conceivable that various projects with different policies share the same artifact server.

I fully agree with this, so I'm not sure whether you've misunderstood my suggestion. Different projects can have their own artifact server configuration (i.e., `artifacts:`, be it via project recommendation or per-project user config) with the same server URL.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 30, 2018, 10:01

Ahh, sorry I missed the hanging *"dict"* word at the end of your first sentence.

Yes I fully agree, configurations for artifact servers are already on a per server basis so it seems like the right approach :)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Dec 7, 2018, 13:48

Just a quick update from my side, I'm currently working on a WIP/PoC on a branch https://gitlab.com/BuildStream/buildstream/tree/tpollard/566 , this approach is as above adding another option to artifact server config, on a per element/project basis. All current tests are passing so default behaviour is preserved, I will shortly be starting on tests to capture the rest of the functionality.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Dec 13, 2018, 17:50

Just a quick update here, I've moved onto testing of the 'partial push' functionality and as expected have hit some painpoints, I'm working through them and should have a WIP available shortly against the branch I listed in the previous comment. One of the biggest issues now is cascache `_reachable_refs_dir()`, which blindly walks through the refs as old behaviour would expect an artifact to be complete quite rightly with no dangling refs. Handling this locally is an annoyance but not too bad, however casserver calls this indirectly, which is a bigger concern to handle as it leads to technically incorrect grpc errors being thrown. For the WIP I think I'm going create/throw a specific exception for a 'dangling' ref dir instead of a generic `filenotfound`, and handle this where appropriate for when the buildtree being missing is acceptable. This implementation is obviously up for debate and it might require the casserver to be less 'dumb' in this regard, however I want to avoid that.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Dec 21, 2018, 15:38

mentioned in merge request !1024

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Dec 21, 2018, 15:46

I've pushed my current WIP state as https://gitlab.com/BuildStream/buildstream/merge_requests/1024 going on the brief agreement on this issue for allowing the configuration option to be set on a per artifact remote config basis. This should allow multiple projects to define the same remote url whilst having their own control of whether to push their respective artifacts with or without a buildtree, as well as the option via user conf.

I've attempted to implement it without explicit reference to 'buildtree' in the cache code, however this has also led to further assumptions being made about where it's acceptable to having missing objects for example in updating mtimes. Reviews at this stage of the implementation would be appreciated, I've included tests to show the behaviour

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 6, 2019, 10:46

marked this issue as related to #896
